File: intellij-plugin/src/main/kotlin/dev/skillbill/intellij/infrastructure/cli/CliGoalStopRepository.kt
package dev.skillbill.intellij.infrastructure.cli

import dev.skillbill.intellij.application.GoalStopOutcome
import dev.skillbill.intellij.application.GoalStopRepository
import dev.skillbill.intellij.application.PreferenceCachePort
import dev.skillbill.intellij.domain.DEFAULT_CLI_TIMEOUT_MS
import dev.skillbill.intellij.domain.DEFAULT_STDERR_LIMIT_BYTES
import dev.skillbill.intellij.domain.DEFAULT_STDOUT_LIMIT_BYTES
import dev.skillbill.intellij.domain.GOAL_STOP_VERB
import dev.skillbill.intellij.domain.REPO_ROOT_OPTION
import java.nio.file.Path

/**
 * CLI-backed [GoalStopRepository]. Runs
 * `skill-bill goal stop <issue-key> --repo-root <canonical>` off the EDT.
 *
 * Runs on a [ProcessRunner] of its own: [ProcessRunner.runCoalesced] coalesces by
 * instance, so sharing the poll runner would let a stop join an in-flight status poll
 * and return that poll's exit code as if the stop had landed.
 */
class CliGoalStopRepository(
    private val preferences: PreferenceCachePort,
    private val processRunner: ProcessRunner,
    private val executableResolver: () -> CliExecutableResolution = {
        CliExecutableResolver.resolve(preferences)
    },
    private val timeoutMs: Long = DEFAULT_CLI_TIMEOUT_MS,
) : GoalStopRepository {
    override suspend fun requestStop(projectRoot: Path, issueKey: String): GoalStopOutcome {
        val key = issueKey.trim()
        if (key.isEmpty()) return GoalStopOutcome.Failed("No issue key to stop")
        val executable = when (val resolution = executableResolver()) {
            is CliExecutableResolution.Found -> resolution.path
            CliExecutableResolution.Missing -> return GoalStopOutcome.Failed("Skill Bill CLI executable not found")
            CliExecutableResolution.Misconfigured ->
                return GoalStopOutcome.Failed("Skill Bill CLI executable override is not usable")
        }
        val canonicalRoot = try {
            projectRoot.toAbsolutePath().normalize().toRealPath()
        } catch (_: Exception) {
            return GoalStopOutcome.Failed("Project root is not a usable path")
        }

        val result = try {
            processRunner.runCoalesced(
                ProcessSpec(
                    command = listOf(executable) + GOAL_STOP_VERB + listOf(key, REPO_ROOT_OPTION, canonicalRoot.toString()),
                    timeoutMs = timeoutMs,
                    stdoutLimitBytes = DEFAULT_STDOUT_LIMIT_BYTES,
                    stderrLimitBytes = DEFAULT_STDERR_LIMIT_BYTES,
                ),
            )
        } catch (_: Exception) {
            // Never leak command output or paths from the failure surface.
            return GoalStopOutcome.Failed("Stop request failed to start")
        }

        return when {
            result.cancelled -> GoalStopOutcome.Failed("Stop request cancelled")
            result.timedOut -> GoalStopOutcome.Failed("Stop request timed out")
            result.exitCode == 0 -> GoalStopOutcome.Requested
            // A non-zero exit is the runtime declining (no live goal, identity mismatch).
            // The reason is not surfaced verbatim; the next poll shows the real state.
            else -> GoalStopOutcome.Failed("Skill Bill declined the stop request")
        }
    }
}

File: vscode-extension/src/infrastructure/cli/CliGoalStopRepository.ts
import * as fs from "fs";
import * as path from "path";
import { GoalStopOutcome, GoalStopRepository } from "../../application/GoalStopRepository";
import { PreferenceCachePort } from "../../application/PreferenceCachePort";
import {
  DEFAULT_CLI_TIMEOUT_MS,
  DEFAULT_STDERR_LIMIT_BYTES,
  DEFAULT_STDOUT_LIMIT_BYTES,
  GOAL_STOP_VERB,
  REPO_ROOT_OPTION,
} from "../../domain/Constants";
import { CliExecutableResolution, resolveCliExecutable } from "./CliExecutableResolver";
import { BoundedProcessResult, ProcessRunner, ProcessSpec } from "./ProcessRunner";

export class CliGoalStopRepository implements GoalStopRepository {
  constructor(
    private readonly preferences: PreferenceCachePort,
    private readonly processRunner: ProcessRunner,
    private readonly executableResolver: () => CliExecutableResolution = () =>
      resolveCliExecutable(this.preferences),
    private readonly timeoutMs: number = DEFAULT_CLI_TIMEOUT_MS,
    private readonly stdoutLimitBytes: number = DEFAULT_STDOUT_LIMIT_BYTES,
    private readonly stderrLimitBytes: number = DEFAULT_STDERR_LIMIT_BYTES,
  ) {}

  async requestStop(projectRoot: string, issueKey: string): Promise<GoalStopOutcome> {
    const key = issueKey.trim();
    if (!key) {
      return { kind: "failed", summary: "No issue key to stop" };
    }
    const resolution = this.executableResolver();
    if (resolution.kind === "missing") {
      return { kind: "failed", summary: "Skill Bill CLI executable not found" };
    }
    if (resolution.kind === "misconfigured") {
      return { kind: "failed", summary: "Skill Bill CLI executable override is not usable" };
    }

    let canonicalRoot: string;
    try {
      canonicalRoot = fs.realpathSync(path.resolve(projectRoot));
    } catch {
      return { kind: "failed", summary: "Project root is not a usable path" };
    }

    let result: BoundedProcessResult;
    try {
      const spec: ProcessSpec = {
        command: [resolution.path, ...GOAL_STOP_VERB, key, REPO_ROOT_OPTION, canonicalRoot],
        timeoutMs: this.timeoutMs,
        stdoutLimitBytes: this.stdoutLimitBytes,
        stderrLimitBytes: this.stderrLimitBytes,
      };
      result = await this.processRunner.runCoalesced(spec);
    } catch {
      return { kind: "failed", summary: "Stop request failed to start" };
    }

    if (result.cancelled) {
      return { kind: "failed", summary: "Stop request cancelled" };
    }
    if (result.timedOut) {
      return { kind: "failed", summary: "Stop request timed out" };
    }
    if (result.exitCode === 0) {
      return { kind: "requested" };
    }
    return { kind: "failed", summary: "Skill Bill declined the stop request" };
  }
}
